### PR TITLE
[RFC][UN-6686] Create an ember cli command to generate translations

### PR DESCRIPTION
**Pull request by raffo0707**
https://uniplaces.atlassian.net/browse/UN-6686

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# EditorConfig helps developers define and maintain consistent
+# coding styles between different editors and IDEs
+# editorconfig.org
+
+root = true
+
```
Comment (montoias) [inline .editorconfig:6]: remove new line

File: .idea/misc.xml [hunk 2]
```diff
@@ -0,0 +1,103 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (montoias) [inline .idea/misc.xml]: What is this?

File: .idea/inspectionProfiles/profiles_settings.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<component name="InspectionProjectProfileManager">
```
Comment (montoias) [inline .idea/inspectionProfiles/profiles_settings.xml]: remove all .idea ugly, scaffolded, auto-generated files :)

File: lib/commands/get.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  _save: function(translations) {
+    var self = this;
+
+    _.each(translations, function(translation, locale) {
+      filesystem.write(filesystem.path(locale), self._content(translation));
+    });
+  },
+
+  _rows: function(tab) {
+    var self = this;
```
Comment (montoias) [inline lib/commands/get.js]: use _.bind

extract 

```
function(res, rej) {
 +      return tab.getRows({offset: 0, limit: tab.rowCount}, function(err, rows) {
 +        if (err) return rej(err);
 +
 +        var translations = self._initTranslations(self.locales, rows);
 +
 +        self._save(translations);
 +
 +        return res(rows);
 +      });
 +    }
```

to a function and do

_.bind(this, this.functionName)

File: lib/commands/get.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        self._save(translations);
+
+        return res(rows);
+      });
+    });
+  },
+
+  run: function (options) {
+    console.log('get translations');
```
Comment (montoias) [inline lib/commands/get.js:69]: keep doing it :)

File: lib/commands/set.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    { name: 'key', type: String, aliases: ['k'] },
+    { name: 'sentence', type: String, aliases: ['s'] },
+    { name: 'default', type: String, aliases: ['d'] }
+  ],
+  locales: {},
+
+  init: function () {
+    var ReadConfigTask = require('../tasks/read-config');
+    var readConfig = new ReadConfigTask({project: this.project});
+    var self = this;
```
Comment (montoias) [inline lib/commands/set.js]: same

File: lib/commands/set.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  },
+
+  _sentence: function(defaultLocale, locale, rawSentence) {
+    return defaultLocale === locale
+      ? rawSentence
+      : spreadsheet.googleTranslateFormula(rawSentence, defaultLocale, locale);
+  },
+
+  _status: function(defaultLocale, locale) {
+    return defaultLocale === locale
```
Comment (montoias) [inline lib/commands/set.js]: one line

File: lib/commands/set.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+      _.setWith(row, locale, sentence, String);
+      _.setWith(row, 'status'+locale, status, String);
+    });
+
+    return row;
+  },
+
+  _set: function(tab, options) {
+    var self = this;
```
Comment (montoias) [inline lib/commands/set.js]: same

File: lib/commands/set.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+        return res(result);
+      });
+    });
+  },
+
+  run: function (options) {
+    console.log('set translations');
+
+    var self = this;
```
Comment (montoias) [inline lib/commands/set.js]: same

File: lib/commands/set.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    var TabSelector = require('../tasks/select-tab.js');
+    var tabSelector = new TabSelector({project: this.project, tabName: options.tab});
```
Comment (montoias) [inline lib/commands/set.js]: if you move  _ember-cli-babel_ in _package.json_ from devDeps to deps, you should have access to all ECMA6 features.

including binds for free with arrow functions.

**Review by montoias — CHANGES_REQUESTED** [review]
remove also .gitkeep

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,36 @@
+---
```
Comment (montoias) [inline .travis.yml]: you can remove this

File: lib/commands/get.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    readConfig.run().then((config) => {
+      if (!config.locales) {
+        throw new SilentError(
+          'Missing loacales array in environment configuration.'
+        );
+      }
+
+      this.keyCol = config.keyCol;
+      config.locales.map((item) => { this.locales[item] = {} });
```
Comment (montoias) [inline lib/commands/get.js]: (item) to item

File: lib/commands/get.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  _content: function(translation) {
+    let prettyTranslsation = JSON.stringify(translation, null, 2);
+
+    return `export default ${prettyTranslsation};`;
+  },
+
+  _initTranslations: function (locales, rows) {
+    let translations = locales;
+
+    rows.map((row) => {
```
Comment (montoias) [inline lib/commands/get.js]: (row) to row

File: lib/commands/get.js [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+      Object.keys(locales).map((key) => {
```
Comment (montoias) [inline lib/commands/get.js]: shouldn't this have a return?

Comment (montoias) [reply lib/commands/get.js]: remove the (key)

File: lib/commands/get.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        translations[key][row[this.keyCol]] = row[key];
```
Comment (montoias) [inline lib/commands/get.js]: make consts out of this, very unreadable :)

File: lib/commands/get.js [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+      });
+    });
+
+    return translations;
+  },
+
+  _save: function(translations) {
+    for (const locale of Object.keys(translations)) {
```
Comment (montoias) [inline lib/commands/get.js]: Object.keys(translations).forEach(locale => x)

File: lib/commands/get.js [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+        return res(rows);
+      });
+    });
+  },
+
+  run: function (options) {
+    console.log('get translations');
+    let TabSelector = require('../tasks/select-tab.js');
+    let tabSelector = new TabSelector(
```
Comment (montoias) [inline lib/commands/get.js]: one line

File: lib/commands/get.js [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+      {
+        project: this.project,
+        tabName: options.tab
+      });
+
+    let tabHandler = tabSelector.run().then((tab) => {
```
Comment (montoias) [inline lib/commands/get.js]: you don't need to return when it's a single line.

you can do

tabSelector.run().then(tab => this._rows(tab));

you can even do something like:

tabSelector.run().then(this._rows);

File: lib/commands/get.js [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+      return this._rows(tab);
+    });
+
+    return tabHandler.then((header) => { return header; });
```
Comment (montoias) [inline lib/commands/get.js]: return tabHandler;

File: lib/commands/set.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+  locales: {},
+  keyCol: '',
+
+  init: function () {
+    let ReadConfigTask = require('../tasks/read-config');
+    let readConfig = new ReadConfigTask({project: this.project});
+
+    readConfig.run().then((config) => {
+      if (!config.locales) {
+        throw new SilentError(
```
Comment (montoias) [inline lib/commands/set.js]: one line

File: lib/commands/set.js [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+  _status: function(defaultLocale, locale) {
+    return defaultLocale === locale ? 'ok' : 'needs attention';
+  },
+
+  _row: function(options) {
+    let row = {[this.keyCol]: options.key};
+
+    this.locales.map((locale) => {
+      row[locale] = this._sentence(options.default, locale, options.sentence);
+      row[`status${locale}`] = this._status(options.default, locale);
```
Comment (montoias) [inline lib/commands/set.js]: return row;

File: lib/commands/set.js [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    });
+
+    return row;
```
Comment (montoias) [inline lib/commands/set.js:53]: remove this

File: lib/commands/set.js [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+  },
+
+  _status: function(defaultLocale, locale) {
+    return defaultLocale === locale ? 'ok' : 'needs attention';
+  },
+
+  _row: function(options) {
+    let row = {[this.keyCol]: options.key};
+
+    this.locales.map((locale) => {
```
Comment (montoias) [inline lib/commands/set.js]: return this.locales....

File: lib/commands/set.js [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+  run: function (options) {
+    console.log('set translations');
+
+    let TabSelector = require('../tasks/select-tab.js');
+    let tabSelector = new TabSelector({
+      project: this.project,
+      tabName: options.tab
+    });
+
+    let tabHandler = tabSelector.run().then((tab) => {
```
Comment (montoias) [inline lib/commands/set.js]: (tab) to tab

File: lib/commands/set.js [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+      return this._set(tab, options);
+    });
+
+    return tabHandler.then((header) => {return header;});
```
Comment (montoias) [inline lib/commands/set.js]: return tabHandler

File: lib/commands/set.js [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+
+    return row;
+  },
+
+  _set: function(tab, options) {
+    return new Promise((res, rej) => {
+      let row = this._row(options);
+
+      return tab.addRow(row, (err, result) => {
+        if (err) return rej(err);
```
Comment (montoias) [inline lib/commands/set.js]: return tab.addRow(row, (err, result) => !err ? res(result) : rej(err));

File: lib/tasks/authenticate.js [hunk 27]
```diff
@@ -0,0 +1,22 @@
+let Task              = require('ember-cli/lib/models/task');
+let Promise           = require('bluebird');
+let GoogleSpreadsheet = require('google-spreadsheet');
+
+module.exports = Task.extend({
+  run: function() {
+    let ReadConfigTask = require('./read-config');
+    let readConfig = new ReadConfigTask({project: this.project});
+
+    let authentication = readConfig.run().then((config) => {
```
Comment (montoias) [inline lib/tasks/authenticate.js]: config

File: lib/tasks/authenticate.js [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+      let spreadsheet = new GoogleSpreadsheet(config.spreadsheetId);
+
+      return new Promise((res) => {
```
Comment (montoias) [inline lib/tasks/authenticate.js]: res

File: lib/tasks/authenticate.js [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+        spreadsheet.useServiceAccountAuth(
+          config.credentials, res.bind(this, spreadsheet)
+        );
+      });
+    });
+
+    return authentication.then((spreadsheet) => {return spreadsheet;});
```
Comment (montoias) [inline lib/tasks/authenticate.js]: return authentication

File: lib/tasks/read-config.js [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+let Promise     = require('ember-cli/lib/ext/promise');
+
+let existsSync  = require('fs').existsSync;
+let path        = require('path');
+let SilentError = require('silent-error');
+
+module.exports = Task.extend({
+  init: function() {
+    if (this._super.init) {
+      this._super.init.apply(this, arguments);
```
Comment (montoias) [inline lib/tasks/read-config.js:11]: this._super.init(...arguments)

File: lib/tasks/read-config.js [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+    if (!this.project) {
+      throw new SilentError('No project passed to read-config task');
+    }
+
+    this.root = this.project.root;
+    this.configPath = this.deployConfigFile || 'config/environment.js';
+    this.superConfigPath = path.resolve(this.root, this.configPath);
+
+    if (!existsSync(this.superConfigPath)) {
+      throw new SilentError(
```
Comment (montoias) [inline lib/tasks/read-config.js]: one line

File: lib/tasks/read-config.js [hunk 32]
```diff
@@ -0,0 +1,42 @@
...
+        `Evironment config does not exist at ${this.superConfigPath}`
+      );
+    }
+  },
+
+  _loadSpreadsheetConfig: function(superConfig) {
+    superConfig._result.credentials = require(
```
Comment (montoias) [inline lib/tasks/read-config.js]: return

File: lib/tasks/read-config.js [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+      `${this.project.root}/${superConfig._result.spreadsheet_cred_file}`
+    );
+
+    return superConfig
```
Comment (montoias) [inline lib/tasks/read-config.js:30]: remove

File: lib/tasks/select-tab.js [hunk 34]
```diff
@@ -0,0 +1,22 @@
...
+          if (err) return rej(err);
+          let tab = info.worksheets.filter((tab) => {
+            return tab.title === this.tabName
+          });
+
+          return res(tab[0])});
+      });
+    });
+
+    return tabHandler.then((tab) => {return tab;});
```
Comment (montoias) [inline lib/tasks/select-tab.js]: return tabHandler

File: lib/tasks/select-tab.js [hunk 35]
```diff
@@ -0,0 +1,22 @@
+let Task = require('ember-cli/lib/models/task');
+
+module.exports = Task.extend({
+  run: function() {
+    let Authenticate = require('./authenticate');
+    let authentication = new Authenticate({project: this.project});
+
+    let tabHandler = authentication.run().then((spreadsheet) => {
```
Comment (montoias) [inline lib/tasks/select-tab.js]: spreadsheet

File: lib/tasks/select-tab.js [hunk 36]
```diff
@@ -0,0 +1,22 @@
...
+      return new Promise((res, rej) =>  {
+        spreadsheet.getInfo((err, info) => {
+          if (err) return rej(err);
+          let tab = info.worksheets.filter((tab) => {
+            return tab.title === this.tabName
+          });
+
+          return res(tab[0])});
```
Comment (montoias) [inline lib/tasks/select-tab.js]: this can never go wrong? 👀

File: lib/utils/filesystem.js [hunk 37]
```diff
@@ -0,0 +1,34 @@
...
+module.exports = {
+  path: function (locale) {
+    return `app/locales/${locale}/translations.js`;
+  },
+
+  dirname: function (filepath) {
+    return path.dirname(filepath);
+  },
+
+  dirSync: function (filepath, permissions) {
```
Comment (montoias) [inline lib/utils/filesystem.js]: permissions = 0777

File: lib/utils/filesystem.js [hunk 38]
```diff
@@ -0,0 +1,34 @@
...
+    if (!filepath) {
+      throw new SilentError('Filepath missing');
+    }
+    if (!permissions) {
```
Comment (montoias) [inline lib/utils/filesystem.js]: remove this

File: lib/utils/filesystem.js (hunk 37 above)
Comment (montoias) [reply lib/utils/filesystem.js]: also create a constant

**Review by montoias — CHANGES_REQUESTED** [review]
remove .gitkeep

File: .editorconfig (hunk 1 above)
Comment (raffo0707) [reply .editorconfig:6]: 👍

**Review by satcin — APPROVED** [review]

**Review by montoias — APPROVED** [review]

**Pull request merged**